### PR TITLE
Allow database character set configuration

### PR DESCRIPTION
**Pull request by dominicchinkh**
Allow database character set configuration

## Description
SuiteCRM administrator can configure "dbconfigoption" as below in config_override.php file so that the MySQL database support storing four bytes per character.

$sugar_config['dbconfigoption']['collation'] = 'utf8mb4_unicode_ci';
$sugar_config['dbconfigoption']['charset'] = 'utf8mb4';

## Motivation and Context
MySQL's encoding called "utf8" (alias of "utf8mb3") only stores a maximum of three bytes per code point. SuiteCRM users often copy and paste text which may include emoji or unicode characters into SuiteCRM text field. This change supports storing supplemental characters in MySQL database. 

## How To Test This
(1) Delete the config.php file. Reinstall the MySQL database
(2) Add a new relationship in Studio

## Types of changes
- [] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Final checklist
- [X] My code follows the code style of this project found [here](https://docs.suitecrm.com/community/contributing-code/coding-standards/).
- [ ] My change requires a change to the documentation.
- [X] I have read the [**How to Contribute**](https://docs.suitecrm.com/community/contributing-code/) guidelines.

### CODE REVIEW

**Comment by connorshea** [comment]
This'd be great to have ❤️

**Comment by samus-aran** [comment]
Hi @dominicchinkh! Thanks for your (multiple) contributions! Can you change the base branch to develop than master as we don't directly merge into master. I'm asking for develop as we are in the midst of a major release so getting it in develop would be quicker than hotfix at the moment.

**Comment by dominicchinkh** [comment]
Base branch change to develop from master

**Comment by dominicchinkh** [comment]
Hi,
I have changed branch to develop from master.

Regards,
Dominic

**Comment by samus-aran** [comment]
Thanks @dominicchinkh! Anyone on the community want to test whilst we review internally?

File: include/database/MysqlManager.php [hunk 1]
```diff
@@ -554,14 +554,20 @@ public function connect(array $configOptions = null, $dieOnError = false)
...
-        mysql_query("SET CHARACTER SET utf8", $this->database);
-        $names = "SET NAMES 'utf8'";
-        $collation = $this->getOption('collation');
-        if (!empty($collation)) {
-            $names .= " COLLATE '$collation'";
+	// cn: using direct calls to prevent this from spamming the Logs
+        $charset = $this->getOption('charset');
+
+        if(!empty($charset)) {
+            mysql_query("SET CHARACTER SET $charset", $this->database);
```
Comment (gymad) [inline include/database/MysqlManager.php]: Please use quote or quoted methods on variable before concatenate it into an sql query.

File: include/database/MysqlManager.php [hunk 2]
```diff
@@ -554,14 +554,20 @@ public function connect(array $configOptions = null, $dieOnError = false)
...
+
+            $names = "SET NAMES '$charset'";
+            $collation = $this->getOption('collation');
+
+            if (!empty($collation)) {
+                $names .= " COLLATE '$collation'";
+            }
+            mysql_query($names, $this->database);
```
Comment (gymad) [inline include/database/MysqlManager.php]: Please use quote or quoted methods on variable before concatenate it into an sql query.

File: include/database/MysqlManager.php [hunk 3]
```diff
@@ -775,12 +781,11 @@ public function createTableSQLParams($tablename, $fieldDefs, $indices, $engine =
...
-        $collation = $this->getOption('collation');
-        if (empty($collation)) {
-            $collation = 'utf8_general_ci';
-        }
-        $sql = "CREATE TABLE $tablename ($columns $keys) CHARACTER SET utf8 COLLATE $collation";
+	// cn: bug 9873 - module tables do not get created in utf8 with assoc collation
+        $collation = $this->getCollation();
+        $charset = $this->getCharset();
+
+        $sql = "CREATE TABLE $tablename ($columns $keys) CHARACTER SET $charset COLLATE $collation";
```
Comment (gymad) [inline include/database/MysqlManager.php]: Please use quote or quoted methods on variable before concatenate it into an sql query.

File: include/database/MysqlManager.php [hunk 4]
```diff
@@ -1506,13 +1548,19 @@ public function createDbUser($database_name, $host_name, $user, $password)
      */
     public function createDatabase($dbname)
     {
-        $this->query("CREATE DATABASE `$dbname` CHARACTER SET utf8 COLLATE utf8_general_ci", true);
+        $collation = $this->getCollation();
+        $charset = $this->getCharset();
+
+        $this->query("CREATE DATABASE `$dbname` CHARACTER SET $charset COLLATE $collation", true);
```
Comment (gymad) [inline include/database/MysqlManager.php]: Please use quote or quoted methods on variable before concatenate it into an sql query.

File: include/database/MysqlManager.php [hunk 5]
```diff
@@ -1506,13 +1548,19 @@ public function createDbUser($database_name, $host_name, $user, $password)
...
     }
 
     public function preInstall()
     {
-        $db->query("ALTER DATABASE `{$setup_db_database_name}` DEFAULT CHARACTER SET utf8", true);
-        $db->query("ALTER DATABASE `{$setup_db_database_name}` DEFAULT COLLATE utf8_general_ci", true);
+        $collation = $this->getCollation();
+        $charset = $this->getCharset();
+
+        $db->query("ALTER DATABASE `{$setup_db_database_name}` DEFAULT CHARACTER SET $charset", true);
```
Comment (gymad) [inline include/database/MysqlManager.php]: Please use quote or quoted methods on variable before concatenate it into an sql query.

File: include/database/MysqlManager.php [hunk 6]
```diff
@@ -1506,13 +1548,19 @@ public function createDbUser($database_name, $host_name, $user, $password)
...
+        $db->query("ALTER DATABASE `{$setup_db_database_name}` DEFAULT COLLATE $collation", true);
```
Comment (gymad) [inline include/database/MysqlManager.php]: Please use quote or quoted methods on variable before concatenate it into an sql query.

File: include/database/MysqliManager.php [hunk 7]
```diff
@@ -348,13 +348,17 @@ public function connect(array $configOptions = null, $dieOnError = false)
...
 
         // cn: using direct calls to prevent this from spamming the Logs
-
         $collation = $this->getOption('collation');
-        if (!empty($collation)) {
-            $names = "SET NAMES 'utf8' COLLATE '$collation'";
+        $charset = $this->getOption('charset');
+
+        if (!empty($collation) && !empty($charset)) {
+            $names = "SET NAMES '$charset' COLLATE '$collation'";
```
Comment (gymad) [inline include/database/MysqliManager.php]: Please use quote or quoted methods on variable before concatenate it into an sql query.

**Comment by samus-aran** [comment]
Hey @dominicchinkh Are you able to review the comments made?

**Comment by dominicchinkh** [comment]
Sorry for late response. Was in a long holiday and busy since then. I have made the requested change and commit. Thank you.

**Comment by Mac-Rae** [comment]
Please resolve the conflicts on this pull request bringing it up to date.

**Comment by dominicchinkh** [comment]
This pull request is ages old and is now out-of-sync with master. I created another pull request under
https://github.com/salesagility/SuiteCRM/pull/7790. Also merge the changes from another related pull request https://github.com/salesagility/SuiteCRM/pull/6475.

**Pull request closed without merging**